setup.py

```diff
@@ -493,38 +573,40 @@ def files(kitten, output, extra_headers=(), extra_sources=(), filter_sources=Non
         files(
             'choose', 'subseq_matcher',
             extra_headers=('kitty/charsets.h',),
-            extra_sources=('kitty/charsets.c',),
+            extra_sources=('kitty/charsets.c',),  # TODO: Compile only once
             filter_sources=lambda x: 'windows_compat.c' not in x),
     ):
-        compile_c_extension(
-            kenv, dest, incremental, compilation_database, all_keys, sources, all_headers + ['kitty/data-types.h'])
+        to_compile.update(prepare_compile_c_extension(
+            kenv, dest, incremental, compilation_database, sources, all_headers + ['kitty/data-types.h']))
+    return to_compile
 
 
 def build(args, native_optimizations=True):
     global env
     try:
-        with open('compile_commands.json') as f:
+        with open('build/compile_commands.json') as f:
             compilation_database = json.load(f)
     except FileNotFoundError:
         compilation_database = []
-    all_keys = set()
     compilation_database = {
         (k['file'], k.get('output')): k['arguments'] for k in compilation_database
     }
     env = init_env(args.debug, args.sanitize, native_optimizations, args.profile, args.extra_logging)
     try:
-        compile_c_extension(
-            kitty_env(), 'kitty/fast_data_types', args.incremental, compilation_database, all_keys, *find_c_files()
-        )
-        compile_glfw(args.incremental, compilation_database, all_keys)
-        compile_kittens(args.incremental, compilation_database, all_keys)
-        for key in set(compilation_database) - all_keys:
-            del compilation_database[key]
```

Comment:
@kovidgoyal could you please explain what the purpose of these two lines is?

Reply:
they remove stale entries from the compilation database. For example if a file was deleted/reamed.